### ISSUE TITLE
Futures?!

### ISSUE THREAD

**Issue by schlamar** [saghul/evergreen#3]
I really dislike the complete future approach in evergreen. Generally the concept is great (compared to callback passing style), but in this case you are mixing two programming models: greenlets vs. futures. 

I would like to see that threads and processes have the same interface than a task. So you would do:

```
t = loop.spawn_[task|thread|process](func, *args, **kwargs)
t.join()
t.result
t.get()  # shortcut to join + result

p = loop.spawn_ext_process(cmd_args, stdout=True)
for line in p.stdout:
    print line
```

Internally you should use libuv's thread pool and process capabilities which should be more efficient than the concurrent.futures implementation.

**Comment by saghul** [comment]
I don't expect everyone to like it. I do, otherwise I wouldn't have done it in the first place :-) Here is the reasoning behind it:

Tasks are cooperative units of work with an API as close as possible to threads. Threads don't store their results, you either store it yourself or use a synchronization mechanism such as a Queue to pass it somewhere. Also, exceptions are just printed. An evergreen task is exactly that, a cooperative thread. And the rest of the API (except futures) is the cooperative equivalent to those synchronization primitives. In a single sentence, Task follows the same model as threads, but it's cooperative.

Given the existence of the GIL, and the fact that I don't and won't provide "green" versions of all modules, deferring operations to threads or processes can come handy, that's where the Futures API comes into play. I adapted it to play nice with tasks, so you can cooperatively wait until some work has been done in a pool of processes or threads.

Evergreen doesn't force you to use either, it gives you the tools so that you can use whatever you like and combine them in a cooperative fashion. This is by design. If you don't like futures, you don't need to use them :-)

Also note that the fact that I used greenlet is just a detail, it's the cooperative model that matters, and I don't see a conflict with blending it with futures.

The API is not written in stone, but I don't think I'll make fundamental changes to the model, I designed it to be exactly what it is, after trying out many things and thinking it through.

You could easily write something like you suggest by hiding different executors, for example.

I'll also be adding a "green" subprocess module when I can find some time to work on that.

**Comment by schlamar** [comment]
OK, I think I expressed my criticism wrong :) I know that you (sometimes) need threads/processes for CPU bound stuff in addition to greenlets/cooperative tasks/multiplexed event loop/...

Let me rephrase: "I don't like how futures are embedded in evergreen". Especially I disagree with `TaskPoolExecutor`. This is basically the same than a `pool.Pool` plus adding some non-cooperative looking API on top on it. This clashes with "there should be only one obvious way to do it" (cooperative tasks, futures, task pool, task pool executor, ?!).

>  I adapted it to play nice with tasks, so you can cooperatively wait until some work has been done in a pool of processes or threads.

This is exactly what it **should** be. `concurrent.futures` patched to work cooperatively without plugging it directly in the core and without some weird and confusing extension (still talking about `TaskPoolExecutor`).

**Comment by saghul** [comment]
Now we are getting somewhere :-)

The idea of having a pool of workers that can do the actual work which you can wait for, use map, etc is what I was after. I find the API in `concurrent.futures` nice for this reason. Now, the futures module in the stdlib has thread and process pool executors, but you may want to use the same API for some i/o bound tasks, so I created the `TaskPoolExecutor`. The whole runner is cooperative, so no matter where you are running the work (a task, a thread, a process) you can cooperatively wait for it, this was the goal with that module.

I agree that there is an overlap between Pool and the TaskPoolExecutor, I'll probably end un removing the former, since it doesn't really do anything that TaskPoolExecutor can't, and doesn't have other nice things like map. Thanks for pointing it out! :-)

Note that you can't patch `cocurrent.futures` without tinkering with the internals, that's why I copied over the code and modified it inplace. Also, I modified the API slightly: https://evergreen.readthedocs.org/en/latest/futures.html#future-class-api-changes I made it a core module which you can choose to use, but it's not used internally (I'm also removing the run_in_executor function from the loop). You can consider it part of the small "evergreen stdlib".

I plan to keep evergreen itself small, that is, providing core functionality which others can extend by writing modules.

**Comment by schlamar** [comment]
> Now we are getting somewhere :-)

Yes :+1: Removing executor stuff from loop and one of (Pool, TaskPoolExecutor) feels right. I would prefer the latter and moving futures to lib/concurrent/futures, so it is accessed just like a patched module (and without your modifications of course). But I could live with the other case, too :)

**Comment by saghul** [comment]
I already removed run_in_executor from the loop and will remove the pool module soon. I'll keep futures where it is, because monkeypatching the stdlib would probably be a nightmare, and I also modified the API, so it's not 100% compatible. This may change in the future (xD) things will probably become clearer after using it for a longer period.

Thanks for the feedback! :-)
